src/main/java/org/jvnet/solaris/libzfs/LibZFS.java

```diff
@@ -47,9 +52,108 @@
 public class LibZFS implements ZFSContainer {
 
     private libzfs_handle_t handle;
+/*
+ * Track features available in current host ZFS ABI so we can use specific JNA
+ * signatures. These will be populated as string tuples (key=value) as defined
+ * in this library privately; beside auto-guesswork they can be populated from
+ * environment variables (via appserver/init-script setup for hosting Jenkins)
+ * as a fallback - but being private, admins should expect to reconfigure such
+ * strings when this library gets updated... in reality, auto-guesswork should
+ * (be implemented at all and then) do a good job.
+ * Note: in this context, the "legacy" ABI refers to function signatures that
+ * were in the OpenSolaris codebase prior to split between Oracle and illumos
+ * and subsequently OpenZFS. The "openzfs" refers to the state of OpenZFS code
+ * "Git HEAD" at the time of relevant edit of these libzfs.jar sources. Keep
+ * in mind that downstream projects may lag accepting such changes, maybe for
+ * years. For this reason, we keep a toggle for every function where we care
+ * about signature differences, and in the (future) case that there are more
+ * than these two levels for us to care about (e.g. HEADs in illumos-gate, ZoL
+ * and *BSD downstreams of OpenZFS) some interim version strings may be defined
+ * and if nothing is defined or value is unknown, then assume "legacy" mode.
+ */
+    private List<Entry<String, String>> libzfs_features;
```

Comment:
I am not sure... one idea about this is that a particular host system has one implementation of ZFS in the kernel (and corresponding libs to load), and I don't think this Java code is instantiated more than once. But if these assumptions seem wrong - I have little idea about how to fix them... did not do Java for over a decade... feel free to improve :-)